contracts/Turnstone.vy

```diff
@@ -0,0 +1,110 @@
+
+struct LogicCallArgs:
+    logicContractAddress: address
+    payload: Bytes[1024]
+
+last_checkpoint: public(bytes32)
+last_valset_id: public(uint256)
+
+@external
+def __init__(turnstone_id: bytes32, validators: DynArray[address, MAX_VALIDATORS], powers: DynArray[uint256, MAX_VALIDATORS]):
+    TURNSTONE_ID = turnstone_id
+    cumulative_power: uint256 = 0
+    i: uint256 = 0
+    for validator in validators:
+        cumulative_power += powers[i]
+        if cumulative_power >= POWER_THRESHOLD:
+            break
+        i += 1
+    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+    valset: Valset = Valset({validators: validators, powers: powers, valset_id: 0})
+    self.last_valset_id = 0
+    new_checkpoint: bytes32 = keccak256(_abi_encode(valset.validators, valset.powers, valset.valset_id, turnstone_id, method_id=method_id("checkpoint(address[],uint256[],uint256,bytes32)")))
+    self.last_checkpoint = new_checkpoint
+
+@external
+@pure
+def turnstone_id() -> bytes32:
+    return TURNSTONE_ID
+
+@internal
+@pure
+def verify_signature(signer: address, hash: bytes32, sig: Signature) -> bool:
+    message_digest: bytes32 = keccak256(concat(convert("\x19Ethereum Signed Message:\n32", Bytes[28]), hash))
+    return signer == ecrecover(message_digest, sig.v, sig.r, sig.s)
+
+@internal
+def check_validator_signatures(consensus: Consensus, hash: bytes32):
+    i: uint256 = 0
+    cumulative_power: uint256 = 0
+    for sig in consensus.signatures:
+        if sig.v != 0:
+            assert self.verify_signature(consensus.valset.validators[i], hash, sig), "Invalid Signature"
+            cumulative_power += consensus.valset.powers[i]
+            if cumulative_power >= POWER_THRESHOLD:
+                break
+        i += 1
+    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+
+@internal
+@view
+def make_checkpoint(valset: Valset) -> bytes32:
+    return keccak256(_abi_encode(valset.validators, valset.powers, valset.valset_id, TURNSTONE_ID, method_id=method_id("checkpoint(address[],uint256[],uint256,bytes32)")))
+
+@external
+def update_valset(new_valset: Valset, consensus: Consensus):
+    assert new_valset.valset_id > consensus.valset.valset_id, "Invalid Valset ID"
+    cumulative_power: uint256 = 0
+    i: uint256 = 0
+    for validator in new_valset.validators:
+        cumulative_power += new_valset.powers[i]
+        if cumulative_power >= POWER_THRESHOLD:
+            break
+        i += 1
+    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+    assert self.last_checkpoint == self.make_checkpoint(consensus.valset), "Incorrect Checkpoint"
+    new_checkpoint: bytes32 = self.make_checkpoint(new_valset)
+    self.check_validator_signatures(consensus, new_checkpoint)
+    self.last_checkpoint = new_checkpoint
+    self.last_valset_id = new_valset.valset_id
+
+@external
+def submit_logic_call(consensus: Consensus, args: LogicCallArgs, message_id: uint256, deadline: uint256):
```

Comment:
This method is vulnerable to a replay attack. I could easily submit the same logic call multiple times. See the invalidation_nonce and invalidation_id code in Gravity.sol for a way to prevent this.

Reply:
Oh, looks like you fixed this with `assert not self.message_id_used[message_id], "Used Message_ID"`